docker/dockerfiles/allin1/services/director/run.sh

```diff
@@ -0,0 +1,248 @@
+column-names=FALSE
+connect-timeout=2
+EOF
+}
+
+exec_sql()
+{
+    mysql --defaults-file=$MYCNF --batch -e "$@"
+}
+
+exec_sql_with_column()
+{
+    mysql --defaults-file=$MYCNF --column-names --batch -e "$@"
+}
+
+exec_sql_with_retry()
+{
+    local sql="$@"
+    while true
+    do
+        result=`exec_sql "$@" 2>&1`
+        ret=$?
+        if [ $ret -eq 0 ] ; then
+            echo "$result"
+            return 0
+        else
+            errcode=`echo $result | awk -F " " '{print $2}'`
+            if [[ $errcode = '1045' || $errcode = '1064' ]] ; then
+                logerror "$PASSWD_ERROR_MSG"
+                return 1
+            else
+                logwarn "mysql command fails with error: '$result', will try again"
+            fi
+        fi
+        sleep $SQL_RETRY_INTERNAL
+    done
+}
+
+check_and_add_be()
+{
+    loginfo "check if need to add BE into FE service ..."
+    while true
+    do
+        result=`exec_sql_with_retry "SHOW BACKENDS;"`
+        ret=$?
+        if [ $ret -ne 0 ] ; then
+            hang_and_die
+        else
+            if echo "$result" | grep -q $MYHOST &>/dev/null ; then
+                loginfo "BE service already added into FE service ... "
+                return 0
+            else
+                beheartbeatport=`be_heartbeat_service_port`
+                loginfo "Add BE($MYHOST:$beheartbeatport) into FE service ..."
+                exec_sql_with_retry "ALTER SYSTEM ADD BACKEND '$MYHOST:$beheartbeatport';"
+            fi
+        fi
+    done
+}
+
+check_and_add_broker()
+{
+    loginfo "check if need to add BROKER into FE service ..."
+    while true
+    do
+        result=`exec_sql_with_retry "SHOW BROKER;"`
+        ret=$?
+        if [ $ret -ne 0 ] ; then
+            hang_and_die
+        else
+            if echo "$result" | grep -q $MYHOST &>/dev/null ; then
+                loginfo "broker service already added into FE service ... "
+                return 0
+            else
+                brokerport=`broker_ipc_port`
+                loginfo "Add BROKER($MYHOST:$brokerport) into FE service ..."
+                exec_sql_with_retry "ALTER SYSTEM ADD BROKER allin1broker '$MYHOST:$brokerport';"
+            fi
+        fi
+    done
+}
+
+loginfo "checking if need to perform auto registring Backend and Broker ..."
+check_fe_fqdn_mismatch
+check_fe_liveness
+generate_my_cnf
+check_and_add_be
+check_and_add_broker
+loginfo "cluster initialization DONE!"
+loginfo "wait a few seconds for BE and Broker's heartbeat ..."
+# allow heartbeat from BE/BROKER to FE
+sleep 10
+loginfo "StarRocks Cluster information details:"
+exec_sql_with_column 'SHOW FRONTENDS\G'
+exec_sql_with_column 'SHOW BACKENDS\G'
+exec_sql_with_column 'SHOW BROKER\G'
+
+loginfo
+loginfo
+loginfo "$PERSISTENT_MSG"
+loginfo
+loginfo "FE mysql query port: `fe_mysql_port`"
+loginfo "FE http service port: 8080"
+loginfo
+loginfo "Enjoy the journal to StarRocks blazing-fast lake-house engine!"
+
+while true
```

Comment:
this `while loop` is used to make run.sh run as a service, to avoid to exit?

Reply:
yes